server/datastore/mysql/operating_systems.go

```diff
@@ -0,0 +1,144 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) ListOperatingSystems(ctx context.Context) ([]fleet.OperatingSystem, error) {
+	return listOperatingSystemsDB(ctx, ds.reader)
+}
+
+func listOperatingSystemsDB(ctx context.Context, tx sqlx.QueryerContext) ([]fleet.OperatingSystem, error) {
+	var os []fleet.OperatingSystem
+	if err := sqlx.SelectContext(ctx, tx, &os, `SELECT * FROM operating_systems`); err != nil {
+		return nil, err
+	}
+	return os, nil
+}
+
+func (ds *Datastore) UpdateHostOperatingSystem(ctx context.Context, hostID uint, hostOS fleet.OperatingSystem) error {
+	os, err := maybeNewOperatingSystemDB(ctx, ds.writer, hostOS)
+	if err != nil {
+		return err
+	}
+	return maybeUpdateHostOperatingSystemDB(ctx, ds.writer, hostID, os.ID)
+}
+
+// maybeNewOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system. If found,
+// it returns the record including the associated ID. If not found, it returns a call
+// to `newOperatingSystemDB`, which inserts a new record and returns the record
+// including the newly associated ID.
+func maybeNewOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	switch os, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return os, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return newOperatingSystemDB(ctx, tx, hostOS)
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get operating system")
+	}
+}
+
+// `newOperatingSystemDB` inserts a record for the given operating system and
+// returns the record including the newly associated ID.
+func newOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	stmt := "INSERT INTO operating_systems (name, version, arch, kernel_version) VALUES (?, ?, ?, ?)"
```

Comment:
`INSERT IGNORE`?

Reply:
@lucasmrod: Sounds good. One thing I noticed is that the id gets auto-incremented in the case of noop `INSERT IGNORE` as well as `INSERT ... ON DUPLICATE KEY UPDATE` so id values will increase more than they need to and we'll see some gaps in the sequence. In this case, we are already checking if the os exists before we call this function so this is really an edge case and it shouldn't have any significant impact on the os id increments. 

In terms of general patterns and practices, have you ever found the "auto-increment on noop `INSERT`" behavior to be an issue in other situations? From what I saw in my googling, it seems people are generally aware of it but don't worry too much about it.